File: ios/Sources/CapacitorVideoKitCore/Gallery.swift
import Foundation
import Photos
import UniformTypeIdentifiers

/**
 A finished video, put where the phone's own gallery will show it.

 The photo library IS the gallery on iOS - there is no folder a file can be written into that
 anything indexes, so a copy into the app's own Documents is a video only this app can open. What
 counts is a `PHAssetCreationRequest`, which is what puts the video in Recents; the album, when one
 is asked for, is a `PHAssetCollection` the new asset is then added to, because an album here holds
 references rather than files.

 `directory` is checked and then has no effect. On Android it chooses between the two folders a
 gallery indexes; iOS has one library and no folders in it, so honouring the option would mean
 inventing a distinction that does not exist. A value Android would refuse is refused here too,
 though, so a typo fails on both platforms rather than only on one.
 */
enum Gallery {

    enum GalleryError: Error {
        /// The person said no, or the library is switched off for this app in Settings.
        case permissionDenied
        /// An option that cannot be honoured: Android's `IllegalArgumentException`, `invalid_spec`
        /// on both.
        case invalidOption(String)
        /// No file at that URL, or nothing that can be read as one.
        case unreadable(String)
        /// The library answered neither with the video nor with an error saying why not.
        case saveFailed(String)
    }

    /**
     The album a save is filed under, or nil for none, once the options have been checked.

     Android's checks, in Android's order and with its words: `directory` first, then `album`. A
     separator in the album name would be a caller asking for a nested album, which the photo library
     cannot express at all, and `.` or `..` is a folder Android would make on disk below API 29 that
     is not a folder of its own - `..` puts the video beside `Movies` - though a photo library would
     take either as a title. Refused rather than flattened, so the same options do the same thing on
     both platforms instead of quietly doing different ones (Android's `Gallery.albumOf`).
     */
    static func album(_ album: String?, directory: String?) throws -> String? {
        switch directory?.lowercased() {
        case nil, "", "movies", "dcim":
            break
        default:
            throw GalleryError.invalidOption("directory is movies or dcim, not: \(directory ?? "")")
        }
        let name = album?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty
        if let name, name.contains("/") || name.contains("\\") || name == "." || name == ".." {
            throw GalleryError.invalidOption("album is one folder name, not a path: \(name)")
        }
        return name
    }

    /**
     Copies `url` into the photo library and answers with the new asset's local identifier, as the
     `ph://` URI the contract documents.

     An identifier rather than a path, which is all iOS ever gives out: the file itself lives inside
     the library, where no app reaches it directly. `album` is what `album(_:directory:)` answered.

     The name is handed to PhotoKit with the file rather than put on a copy of it:
     `PHAssetResourceCreationOptions.originalFilename` is exactly "call it this", and a whole second
     video in a temporary folder is what a nearly full phone would fail to write.
     */
    static func save(url: URL, fileName: String?, album: String?) async throws -> String {
        guard FileManager.default.isReadableFile(atPath: url.path) else {
            throw GalleryError.unreadable("there is no file to read at \(url.path)")
        }

        let mayFile = try await requestPermission(album: album != nil)
        let collection = mayFile ? album.flatMap { albumCollection(named: $0) } : nil

        let options = PHAssetResourceCreationOptions()
        // Copied, never moved: the source is the app's own render and the caller may still want it.
        options.shouldMoveFile = false
        if let name = fileName?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty {
            options.originalFilename = name
            // PhotoKit reads the type off the file's extension. A file a host kept without one
            // takes it from the name it is being saved under instead.
            let ext = (name as NSString).pathExtension
            if url.pathExtension.isEmpty, !ext.isEmpty, let type = UTType(filenameExtension: ext) {
                options.uniformTypeIdentifier = type.identifier
            }
        }

        return try await withCheckedThrowingContinuation { continuation in
            var identifier: String?
            PHPhotoLibrary.shared().performChanges {
                let request = PHAssetCreationRequest.forAsset()
                request.addResource(with: .video, fileURL: url, options: options)
                guard let placeholder = request.placeholderForCreatedAsset else { return }
                identifier = placeholder.localIdentifier

                if let collection, let add = PHAssetCollectionChangeRequest(for: collection) {
                    add.addAssets([placeholder] as NSArray)
                }
            } completionHandler: { done, error in
                if let error {
                    // As it came, so `rejection(for:)` can read the space and access codes off it.
                    continuation.resume(throwing: error)
                } else if done, let identifier {
                    continuation.resume(returning: "ph://\(identifier)")
                } else {
                    continuation.resume(throwing: GalleryError.saveFailed("the photo library refused the video"))
                }
            }
        }
    }

    /**
     The message and `code` a failed save is rejected with.

     Always one of `SaveToGalleryFailureCode`, plus the `invalid_spec` Android answers an option it
     cannot honour with - never `io`, which a host switching on that union has no case for. A full
     disk is `no_space` wherever it surfaces, so the host can say "free up space" rather than
     "try again"; a file PhotoKit cannot read as a video is `unreadable_input`, as a missing one is;
     and whatever else the library says is `unknown`, with its words as the message.
     */
    static func rejection(for error: Error) -> (message: String, code: String) {
        switch error {
        case GalleryError.permissionDenied:
            return ("The photo library is not available to this app", Reject.permissionDenied)
        case let GalleryError.invalidOption(message):
            return (message, Reject.invalidSpec)
        case let GalleryError.unreadable(message):
            return (message, Reject.unreadableInput)
        case let GalleryError.saveFailed(message):
            return (message, ComposeFailureCode.unknown.rawValue)
        default:
            break
        }

        let message = ErrorMapping.describe(error)
        if ErrorMapping.isOutOfSpace(error) { return (message, Reject.noSpace) }
        let nsError = error as NSError
        if nsError.domain == PHPhotosErrorDomain {
            switch PHPhotosError.Code(rawValue: nsError.code) {
            case .notEnoughSpace:
                return (message, Reject.noSpace)
            case .accessUserDenied, .accessRestricted:
                return (message, Reject.permissionDenied)
            case .invalidResource, .missingResource:
                return (message, Reject.unreadableInput)
            default:
                break
            }
        }
        return (message, ComposeFailureCode.unknown.rawValue)
    }

    /// What a save that was asked for an album does about it, given the read/write access held.
    enum AlbumStep: Equatable {
        /// Ask for read/write access before saving; the one prompt settles adding too.
        case ask
        /// File the video into the album.
        case file
        /// Save it to Recents and leave the album out.
        case skip
    }

    /**
     Whether an album can be had, and what it takes.

     Finding an album and making one both need read/write access, which add-only access does not
     include, so with add-only access PhotoKit refuses both and the album would be dropped without
     a word. So when read/write access has never been asked about, THAT is what is asked.

     Only when the host is allowed to ask, though: iOS terminates an app that asks for read/write
     access without `NSPhotoLibraryUsageDescription` in its `Info.plist`, and the README tells a
     host that only saves that `NSPhotoLibraryAddUsageDescription` is all it needs. Such a host's
     saves go to Recents with the album left out, rather than taking the app down on the first one.

     Only full access files into the album. With add-only access there is no album to be had; with a
     limited grant PhotoKit does not promise that an album made on an earlier save is visible, and
     not seeing it would make another album of the same name on every save. Either way the video is
     still saved, in Recents, and skipping the album on purpose beats a shelf of duplicates.
     */
    static func albumStep(readWrite: PHAuthorizationStatus, hostDeclaresReading: Bool) -> AlbumStep {
        guard hostDeclaresReading else { return .skip }
        switch readWrite {
        case .notDetermined: return .ask
        case .authorized: return .file
        default: return .skip
        }
    }

    /**
     Permission to add, and whether the video may be filed in an album as well (`albumStep`).

     `.addOnly` is all the save itself needs, and asking for `.readWrite` for that alone would put
     a prompt about every photo they own in front of the person in exchange for writing one video.
     With an album it is the one prompt, and it has a price: photo library access is one setting
     per app, so a person who answers it with Don't Allow may have refused adding along with
     reading, and then the save is refused with `permission_denied` where an add-only prompt might
     have been allowed.
     */
    private static func requestPermission(album: Bool) async throws -> Bool {
        let declared = hostDeclaresReading
        if album, albumStep(readWrite: PHPhotoLibrary.authorizationStatus(for: .readWrite),
                            hostDeclaresReading: declared) == .ask {
            _ = await PHPhotoLibrary.requestAuthorization(for: .readWrite)
        }
        try await requestAddPermission()
        return album && albumStep(readWrite: PHPhotoLibrary.authorizationStatus(for: .readWrite),
                                  hostDeclaresReading: declared) == .file
    }

    /// Whether the host's `Info.plist` says why it reads the photo library, which is what lets it
    /// ask to at all.
    private static var hostDeclaresReading: Bool {
        Bundle.main.object(forInfoDictionaryKey: "NSPhotoLibraryUsageDescription") != nil
    }

    /**
     Permission to ADD, which is narrower than permission to read, asked for when it has not been
     settled - which after a read/write prompt it usually already has.
     */
    private static func requestAddPermission() async throws {
        let current = PHPhotoLibrary.authorizationStatus(for: .addOnly)
        if current == .authorized || current == .limited { return }
        if current == .denied || current == .restricted { throw GalleryError.permissionDenied }

        let granted = await withCheckedContinuation { continuation in
            PHPhotoLibrary.requestAuthorization(for: .addOnly) { continuation.resume(returning: $0) }
        }
        guard granted == .authorized || granted == .limited else { throw GalleryError.permissionDenied }
    }

    /**
     The album by that name, made when there is not one yet.

     Nil when the library refused to make one, and nil is not a failure: the video is still saved,
     it is simply in Recents and nowhere else. An album is where a save is filed, not whether it
     happened, and losing the whole video over a folder would be the wrong trade.
     */
    private static func albumCollection(named name: String) -> PHAssetCollection? {
        let options = PHFetchOptions()
        options.predicate = NSPredicate(format: "localizedTitle = %@", name)
        let existing = PHAssetCollection.fetchAssetCollections(with: .album, subtype: .any, options: options)
        if let found = existing.firstObject { return found }

        var identifier: String?
        try? PHPhotoLibrary.shared().performChangesAndWait {
            let request = PHAssetCollectionChangeRequest.creationRequestForAssetCollection(withTitle: name)
            identifier = request.placeholderForCreatedAssetCollection.localIdentifier
        }
        guard let identifier else { return nil }
        return PHAssetCollection.fetchAssetCollections(
            withLocalIdentifiers: [identifier], options: nil
        ).firstObject
    }
}

private extension String {
    var nilIfEmpty: String? { isEmpty ? nil : self }
}

File: android/src/main/java/net/dotnetdreamer/videokit/videocomposer/Gallery.kt
package net.dotnetdreamer.videokit.videocomposer

import android.content.ContentValues
import android.content.Context
import android.media.MediaScannerConnection
import android.net.Uri
import android.os.Build
import android.os.Environment
import android.provider.MediaStore
import android.webkit.MimeTypeMap
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.util.Locale

/**
 * A finished video, put where the phone's own gallery will show it.
 *
 * The two obvious ways to do this are both wrong, and both were in use before this existed:
 *
 * - Copying into `getExternalMediaDirs()` - `Android/media/<package>/` - puts the file in the app's
 *   OWN media storage. Android deletes that directory when the app is uninstalled, so every video
 *   a customer ever "saved to their gallery" goes with the app.
 * - Announcing the copy with an `ACTION_MEDIA_SCANNER_SCAN_FILE` broadcast. Deprecated at API 29
 *   and a no-op on the releases after it, so on a newer phone nothing is ever told the file exists.
 *
 * A MediaStore insert is neither: the row IS the announcement, the file belongs to the device
 * rather than to this app, and from API 29 it needs no permission because the insert is scoped to
 * the one directory it names. Below 29 there is no scoped insert, so the file is written into the
 * public directory and handed to the scanner through [MediaScannerConnection] - which, unlike the
 * broadcast, is still supported.
 */
object Gallery {

    /** What a video is when its own name does not say. Every render this package makes is one. */
    private const val DEFAULT_MIME_TYPE = "video/mp4"

    /** What a video is called when neither the caller nor its source names it. */
    private const val DEFAULT_NAME = "video.mp4"

    /**
     * Copies `uri` into the gallery and answers with the row it now occupies.
     *
     * Throws [IllegalArgumentException] for an option that cannot be honoured and [IOException] for
     * a copy that did not finish; the plugin turns those into the codes `definitions.ts` promises.
     */
    fun save(
        context: Context,
        uri: String,
        fileName: String?,
        album: String?,
        directory: String?,
    ): Uri {
        val source = Uri.parse(uri).let { if (it.scheme == null) Uri.fromFile(File(uri)) else it }
        val name = nameOf(fileName, source.lastPathSegment)
        val folder = folderOf(directory)
        val subFolder = albumOf(album)

        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            insert(context, source, name, folder, subFolder)
        } else {
            writeToPublicDirectory(context, source, name, folder, subFolder)
        }
    }

    /**
     * The gallery's own copy, made by MediaStore rather than by this package.
     *
     * Inserted pending and cleared at the end, which is what keeps a gallery from showing a
     * half-copied video: while `IS_PENDING` is set the row is invisible to every other app. A copy
     * that fails takes its row with it for the same reason - a row with no bytes behind it is a
     * gallery entry that plays nothing.
     */
    private fun insert(context: Context, source: Uri, name: String, folder: String, album: String?): Uri {
        val resolver = context.contentResolver
        val pending = ContentValues().apply {
            put(MediaStore.Video.Media.DISPLAY_NAME, name)
            put(MediaStore.Video.Media.MIME_TYPE, mimeTypeOf(name))
            put(
                MediaStore.Video.Media.RELATIVE_PATH,
                if (album == null) folder else "$folder${File.separator}$album",
            )
            put(MediaStore.Video.Media.IS_PENDING, 1)
        }

        val item = resolver.insert(
            MediaStore.Video.Media.getContentUri(MediaStore.VOLUME_EXTERNAL_PRIMARY),
            pending,
        ) ?: throw IOException("the gallery gave nowhere to put it")

        try {
            resolver.openInputStream(source).use { input ->
                input ?: throw IOException("there is nothing to read at $source")
                resolver.openOutputStream(item).use { output ->
                    output ?: throw IOException("the gallery gave nothing to write to")
                    input.copyTo(output, DEFAULT_BUFFER_SIZE)
                }
            }
        } catch (e: Throwable) {
            resolver.delete(item, null, null)
            throw e
        }

        resolver.update(item, ContentValues().apply { put(MediaStore.Video.Media.IS_PENDING, 0) }, null, null)
        return item
    }

    /** The same video on a phone with no scoped insert, where a gallery is a directory on disk. */
    private fun writeToPublicDirectory(
        context: Context,
        source: Uri,
        name: String,
        folder: String,
        album: String?,
    ): Uri {
        val root = Environment.getExternalStoragePublicDirectory(folder)
        val target = if (album == null) root else File(root, album)
        if (!target.isDirectory && !target.mkdirs()) throw IOException("could not make $target")

        val file = File(target, name)
        context.contentResolver.openInputStream(source).use { input ->
            input ?: throw IOException("there is nothing to read at $source")
            FileOutputStream(file).use { output -> input.copyTo(output, DEFAULT_BUFFER_SIZE) }
        }

        MediaScannerConnection.scanFile(context, arrayOf(file.absolutePath), arrayOf(mimeTypeOf(name)), null)
        return Uri.fromFile(file)
    }

    /**
     * The album, trimmed, or null for none; refused when it is not one folder name.
     *
     * A separator here would be a caller asking for a nested album, which iOS cannot express at
     * all - a photo library has albums and no folders under them. `.` and `..` are no folder of
     * their own either: below API 29 the album is a directory made on disk, and `..` would put the
     * video beside `Movies` rather than in it. Refused rather than flattened, so the same options
     * do the same thing on both platforms instead of quietly doing different ones; iOS's
     * `Gallery.album` refuses the same names in the same words.
     */
    internal fun albumOf(album: String?): String? {
        val name = album?.trim()?.takeIf { it.isNotEmpty() } ?: return null
        require(!name.contains('/') && !name.contains('\\') && name != "." && name != "..") {
            "album is one folder name, not a path: $name"
        }
        return name
    }

    /**
     * The name the video is saved under: [fileName], or else the source's own name, or else
     * `video.mp4`, trimmed, and made a NAME rather than a path.
     *
     * Below API 29 the video is written with `File(directory, name)`, so a name with a separator in
     * it - `a/../../x.mp4` - would put it outside `Movies` or `DCIM`, anywhere on shared storage.
     * From 29 on, MediaStore makes a valid file name of `DISPLAY_NAME` itself, a separator becoming
     * `_`, and files the row by `RELATIVE_PATH` alone. So a separator becomes `_` here, on every
     * release, and a name that is `.` or `..`, which names a folder rather than a file, is
     * `video.mp4` as a missing one is. Flattened rather than refused, unlike [albumOf]: a name is
     * what the gallery prints under the video, and a slash in a title is no reason to lose a save.
     */
    internal fun nameOf(fileName: String?, sourceName: String?): String {
        val name = (fileName?.takeIf { it.isNotBlank() } ?: sourceName ?: DEFAULT_NAME).trim()
            .replace('/', '_')
            .replace('\\', '_')
        return if (name.isEmpty() || name == "." || name == "..") DEFAULT_NAME else name
    }

    /**
     * `movies` and `dcim` and nothing else, because those are the two a gallery indexes and the
     * contract says so. An unknown value is the caller's mistake rather than a reason to guess.
     */
    private fun folderOf(directory: String?): String = when (directory?.lowercase(Locale.ROOT)) {
        null, "", "movies" -> Environment.DIRECTORY_MOVIES
        "dcim" -> Environment.DIRECTORY_DCIM
        else -> throw IllegalArgumentException("directory is movies or dcim, not: $directory")
    }

    /**
     * What the container is, read off the name the video is being saved under.
     *
     * MediaStore checks the extension against the type and renames the file when the two disagree,
     * so guessing `video/mp4` for a `.mov` would file somebody's clip as `holiday.mov.mp4`.
     */
    private fun mimeTypeOf(fileName: String): String {
        val extension = fileName.substringAfterLast('.', "")
        if (extension.isEmpty()) return DEFAULT_MIME_TYPE
        return MimeTypeMap.getSingleton().getMimeTypeFromExtension(extension.lowercase(Locale.ROOT))
            ?: DEFAULT_MIME_TYPE
    }
}
